proxy/common.c

```diff
@@ -372,16 +427,16 @@ gridd_request_replicated (struct req_args_s *args, struct client_ctx_s *ctx,
 		g_ptr_array_add (bodyv, body);
 		g_ptr_array_add (urlv, g_strdup(url));
 
-		/* Check for a possible redirection */
-		if (flag_prefer_master_for_read || flag_prefer_slave_for_read
-				|| flag_prefer_master_for_write) {
+		/* If there is no error that indicate a wrong cache entry, we can
+		 * check for a possible redirection to update the cache of master */
+		if ((!err || !error_clue_for_decache(err))
+				&& (flag_prefer_master_for_read ||
+					flag_prefer_slave_for_read ||
+					flag_prefer_master_for_write)) {
 			const char *actual = client ? gridd_client_url(client) : NULL;
-			// if (actual && 0 != strcmp(actual, _id)) {
 			if (actual && 0 != strcmp(actual, url)) {
-				gchar *k = g_strdup(election_key);
-				gchar *v = g_strdup(actual);
-				GRID_TRACE("MASTER %s %s", v, k);
-				MASTER_WRITE(lru_tree_insert(srv_master, k, v));
+				// TODO(jfs): maybe save the ID instead (or remove this mark)
+				service_learn_master(election_key, actual);
```

Comment:
If `actual` is "the service that responded without error", and `flag_prefer_slave_for_read` is set, then we are filling the cache with the wrong service.

Reply:
`actual` is the service we have been redirected to. The only redirection case is the need for a master, the retry with another peer is a piece of logic not managed by the gridd_client API (but by the proxy). So we could save any `actual` that is different from the initial target.

The check on the flags is just here to save a few cycles if possible: if you never use any preference, then you don't need to keep cached information. If the management of preferences becomes request-dependant, the check on the flags must even be removed.